workers/task-driver/src/driver.rs

```diff
@@ -281,121 +247,89 @@ impl TaskExecutor {
 
     /// Start the given task, preempting the queues of the given wallets
     async fn start_preemptive_task(
+        &self,
         wallet_ids: Vec<WalletIdentifier>,
         task_id: TaskIdentifier,
         task: TaskDescriptor,
-        state: State,
-        preemptive_tasks: Shared<HashSet<TaskIdentifier>>,
-        task_fut: impl Future<Output = Result<(), TaskDriverError>>,
     ) -> Result<(), TaskDriverError> {
         // Pause the queues for the affected local wallets
         for wallet_id in wallet_ids.iter() {
-            state.pause_task_queue(wallet_id, task_id, task.clone())?.await?;
+            self.state().pause_task_queue(wallet_id, task_id, task.clone()).await?;
         }
 
         // Add the task to the preemptive tasks list so that notification requests can
         // be registered for it
-        preemptive_tasks.write().unwrap().insert(task_id);
+        self.preemptive_tasks.write().unwrap().insert(task_id);
 
-        let res = task_fut.await;
+        let res = self.start_task(true /* immediate */, task_id, task).await;
         if let Err(e) = &res {
             error!("error running immediate task: {e:?}");
         }
 
         // Unpause the queues for the affected local wallets
         for wallet_id in wallet_ids.iter() {
-            state.resume_task_queue(wallet_id, res.is_ok())?;
+            self.state().resume_task_queue(wallet_id, res.is_ok()).await?;
         }
 
         // Remove from the preemptive tasks list
-        preemptive_tasks.write().unwrap().remove(&task_id);
-
+        self.preemptive_tasks.write().unwrap().remove(&task_id);
         Ok(())
     }
 
-    /// Get the future for a task execution
-    fn create_task_future(
-        &self,
-        immediate: bool,
-        task_id: TaskIdentifier,
-        descriptor: TaskDescriptor,
-    ) -> impl Future<Output = Result<(), TaskDriverError>> {
-        // Collect the arguments then spawn
-        let ctx = self.task_context();
-        let args = self.runtime_config;
-        let task_notifications = self.task_notifications.clone();
-
-        Self::start_task(immediate, task_id, descriptor, ctx, args, task_notifications)
-    }
-
     /// Spawn a new task in the driver
     ///
     /// Returns the success of the task
     async fn start_task(
```

Comment:
can we re-attach the tracing that was lost from the `TaskDriverJob::Run` arm of the match statement in `run` to `start_task` here? Namely:
```suggestion
    #[instrument(name = "task", skip_all, err, fields(
        task_id = %id,
        task = %task.display_description(),
        queue_key = %task.queue_key(),
    )]
    async fn start_task(
```
this is by far the most important trace we produce. And applying it this way to `start_task` will also give us more descriptive traces for preemptive tasks, though I would also request that we add a similar `#[instrument]` attribute to `start_preemptive_task` so that we can trace any errors in pausing/unpausing the queue appropriately.

Reply:
done